### PR TITLE
Add Autograd Support for Low-Level Bessel Functions

### PR DESCRIPTION
**Pull request by colalb1**
Makes the eight low-level Bessel special functions differentiable (backward + forward-mode AD), replacing their `non_differentiable` autograd entries.

Changes:
- `derivatives.yaml`: `auto_element_wise` derivative formulas for `special_bessel_{j0,j1,y0,y1}` and `special_modified_bessel_{i0,i1,k0,k1}`
- `FunctionsManual.{cpp,h}`: new `bessel_j1_backward` handling the `0/0` removable singularity of `J1'` at `x = 0` (limit `1/2`); `I1` reuses the existing `i1_backward`
- OpInfo (`special.py`): enable `supports_forward_ad`, `supports_fwgrad_bwgrad`, and `backward_dtypes` so the ops get gradcheck/gradgradcheck/forward-AD and SciPy reference coverage automatically.
- `test_autograd.py`: add `TestBesselAutograd.test_zero_limit`, the one check OpInfo cannot express since gradcheck cannot differentiate `0/0` at `x=0`

**Questions:**
- `test_zero_limit` currently lives in `test_autograd.py`. Should it move to its own file instead of sitting alongside the autograd engine tests?
- The CPU forward Bessel kernels are only ~1e-7 accurate in float64, which constrains gradient tolerances. Should improving the CPU Bessel ops be a hard blocker for this PR, and should I implement that first before moving forward here?

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -6121,6 +6121,33 @@ Tensor i1_backward(
...
+      self.scalar_type(), "bessel_j1_backward", [&]() {
+        // J1'(x) = J0(x) - J1(x) / x. At x = 0 this is 0/0, but the
+        // analytic limit is 0.5, so we clamp it for near-zero inputs. The
+        // `where` on a `safe_self` keeps NaNs from the untaken branch out of
+        // the gradient.
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
+        auto self_is_not_tiny = self.abs() > eps;
+
+        auto safe_self = at::where(
+            self_is_not_tiny, self, at::scalar_tensor(eps, self.options()));
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Should use masked_fill instead?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -6121,6 +6121,33 @@ Tensor i1_backward(
...
+        auto gradx =
+            (at::special_bessel_j0(safe_self) -
+             (result * safe_self.reciprocal()));
+             
+        return grad *
+            at::where(
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:6375]: Also should be masked_fill?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: Addressed.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:6375]: Addressed.

File: torch/testing/_internal/opinfo/definitions/special.py [hunk 3]
```diff
@@ -121,6 +121,20 @@ def sample_inputs_erfcx(op_info, device, dtype, requires_grad, **kwargs):
 _unsigned_int_types = (torch.uint16, torch.uint32, torch.uint64)
 
 
+# Restricting the x > 0 Bessel ops to domain=(0, None) makes the unary sample
+# generator clamp samples with clamp_min, which has no CPU kernel for the barebones
+# unsigned integer types. Skip the affected forward reference-numerics tests there.
+_domain_uint_skips = tuple(
+    DecorateInfo(
+        unittest.skip("clamp_min unimplemented for unsigned int types"),
```
Comment (Skylion007) [inline torch/testing/_internal/opinfo/definitions/special.py]: Okay, this should DEFINITELY be an xfail

Comment (Skylion007) [reply torch/testing/_internal/opinfo/definitions/special.py]: Also we should just implement clamp_min, it's easy enough to implement, probably just need a few more dispatches

Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/special.py]: Changed this to an xfail. Will make a different PR for clamp_min.

Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/special.py]: @Skylion007 PR for adding uints to the clamp functions on CPU and CUDA: https://github.com/pytorch/pytorch/pull/189953

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Ideally add the missing dtype coverage to the clamp tests in the other PR, then remove the xfails here. :)

File: torch/testing/_internal/opinfo/definitions/special.py [hunk 4]
```diff
@@ -730,7 +744,10 @@ def sample_inputs_erfcx(op_info, device, dtype, requires_grad, **kwargs):
         dtypes=all_types_and(torch.bool, *_unsigned_int_types),
         dtypesIfMPS=all_types_and(torch.bool, torch.float16, torch.bfloat16),
         ref=scipy.special.k0 if TEST_SCIPY else None,
-        supports_autograd=False,
+        backward_dtypes=floating_types(),
+        domain=(0, None),
```
Comment (Skylion007) [inline torch/testing/_internal/opinfo/definitions/special.py:775]: Huh, didn't know about the domain arg?

File: torch/testing/_internal/opinfo/definitions/special.py [hunk 5]
```diff
@@ -501,7 +508,10 @@ def sample_inputs_erfcx(op_info, device, dtype, requires_grad, **kwargs):
         dtypes=all_types_and(torch.bool, *_unsigned_int_types),
         dtypesIfMPS=all_types_and(torch.bool, torch.float16, torch.bfloat16),
         ref=scipy.special.y1 if TEST_SCIPY else None,
-        supports_autograd=False,
+        backward_dtypes=floating_types(),
+        domain=(0, None),
```
Comment (Skylion007) [inline torch/testing/_internal/opinfo/definitions/special.py:536]: Why do you need domain now, but not before?

**Comment by Skylion007** [comment]
@claude review this and double check the op tests are correct.

File: torch/testing/_internal/opinfo/definitions/special.py (hunk 5 above)
Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/special.py:536]: @Skylion007  I will look into this in the next couple of days and get back to you with a thorough explanation.

Could you review https://github.com/pytorch/pytorch/pull/189953 when you get the chance? This is the blocking clamp fix for this PR, just want to make sure this gets in before everything in this is addressed.

Thank you.

File: test/test_unary_ufuncs.py [hunk 6]
```diff
@@ -1207,6 +1207,20 @@ def test_sinc(self, device, dtype):
...
+    # bessel_j1 and modified_bessel_i1 have a 0/0 closed-form gradient at x = 0
+    # whose analytic limit is 1/2. OpInfo sample generation does not reliably emit
+    # exact zeros, so this special-cased value is not covered there; check it here.
+    @parametrize(
+        "name",
+        ("bessel_j1", "modified_bessel_i1"),
+    )
+    def test_bessel_zero_limit_gradient(self, device, name):
+        op = getattr(torch.special, name)
+        x = torch.zeros(4, dtype=torch.double, device=device, requires_grad=True)
```
Comment (albanD) [inline test/test_unary_ufuncs.py]: Won't the torch.double fail on MPS?
 @dtypes(torch.double) like the above would fix this.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -6121,6 +6121,28 @@ Tensor i1_backward(
...
+  return AT_DISPATCH_FLOATING_TYPES(
+      self.scalar_type(), "bessel_j1_backward", [&]() {
+        // J1'(x) = J0(x) - J1(x) / x. At x = 0 this is 0/0, but the analytic
+        // limit is 0.5. Fill the singular points with a safe input so the
+        // reciprocal stays finite (cf. i1_backward, #52248), then fill the
+        // resulting gradient there with the 0.5 limit.
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
+        auto self_is_tiny = self.abs() <= eps;
+
+        auto safe_self = self.masked_fill(self_is_tiny, eps);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp]: Why masked_fill and not where like the above?

File: torch/testing/_internal/opinfo/definitions/special.py (hunk 5 above)
Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/special.py:536]: @Skylion007 This addresses both comments about the domain args:

The four functions `special.bessel_y0`, `special.bessel_y1`, `special.modified_bessel_k0`, `special.modified_bessel_k1` are only real-valued and finite for $x > 0$ since they all have singularities at $0$ and are complex/undefined for $x < 0$. The other Bessel functions (`bessel_j0/j1`, `modified_bessel_i0/i1`) are defined on the whole real line, so they require no domain restriction.

The domain restrictions are required now because `supports_autograd` is flipped to True in this PR, so before this, only forward-numeric tests ran. Forward numeric tests tolerate negative inputs because `torch` and `scipy` both return NaN there and the comparison treats NaN == NaN as equal. `gradcheck`/`gradgradcheck` draw sample inputs and perturb them with finite differences, so on negative or near-zero inputs, $Y_n/ K_n$ and their derivatives blow up to NaN/inf, so `gradcheck` would fail. `domain=(0, None)` keeps the samples in the region where the function and its derivatives are well-defined.

Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/special.py:536]: [Chapter 10](https://dlmf.nist.gov/10) should support the math-related statements above

File: test/test_unary_ufuncs.py (hunk 6 above)
Comment (colalb1) [reply test/test_unary_ufuncs.py]: Yes. Addressed.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 7 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: I used `masked_fill()` instead of `where(~mask,...)` because it avoids two-gradient branching that `where` contains. Also, I find it a bit more readable.

I will change it to the `where` function since this matches other local functions more closely.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (Skylion007) [reply torch/csrc/autograd/FunctionsManual.cpp]: Nah, it hasn't Any reason we still use at::where?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (Skylion007) [reply torch/csrc/autograd/FunctionsManual.cpp:6375]: how so?

**Comment by Skylion007** [comment]
Ah nvm I see we should do a global replace with masked_fill later

**Comment by Skylion007** [comment]
Actually at where is better here

**Comment by Skylion007** [comment]
@claude run a check focusing on the correctness and numerical stability of the backward passes added here.

**Comment by Skylion007** [comment]
@claude just verify the numerical stabilities of the derivatives

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 8]
```diff
@@ -6179,6 +6179,32 @@ Tensor i1_backward(
...
+Tensor bessel_j1_backward(
+    const Tensor& grad,
+    const Tensor& self,
+    const Tensor& result) {
+  return AT_DISPATCH_FLOATING_TYPES(
+      self.scalar_type(), "bessel_j1_backward", [&]() {
+        // J1'(x) = J0(x) - J1(x) / x. At x = 0 this is 0/0, but the analytic
+        // limit is 0.5. Replace the singular points with a safe input so the
+        // reciprocal stays finite, then select the 0.5 limit there.
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Slight nit can be constexpr I guess but NBD

**Review by Skylion007 — APPROVED** [review]

**Comment by colalb1** [comment]
@Skylion007 Which branch would be appropriate to merge this into given the [dependent clamp PR](https://github.com/pytorch/pytorch/pull/189953) was merged?

**Comment by Skylion007** [comment]
Main branch is good

**Comment by Skylion007** [comment]
There are a few minor issues with NaN handling the Claude pointed out that would be great to resolve.

**Comment by colalb1** [comment]
> There are a few minor issues with NaN handling the Claude pointed out that would be great to resolve.

@Skylion007 
I had claude fix the issues that other claude found. I also added the `constexpr` to the `eps` that you raised earlier. The branch is up-to-date with main.

If there are any other changes I should make, please let me know.

File: test/test_unary_ufuncs.py [hunk 9]
```diff
@@ -1207,6 +1207,38 @@ def test_sinc(self, device, dtype):
         )
         gradcheck(torch.sinc, a)
 
+    # The order-1 Bessel gradients are indeterminate at x = 0 as written: 0/0 for
+    # j1/i1 (limit 1/2) and (-inf) - (-inf) for y1 (limit +inf). OpInfo sample
+    # generation does not reliably emit exact zeros, and y1's domain floors samples
+    # away from 0, so these special-cased values are only covered here.
+    @dtypes(torch.double)
```
Comment (Skylion007) [inline test/test_unary_ufuncs.py]: This is going to fail on MPS because it doesn't have FP64 on the accelerator.

Comment (colalb1) [reply test/test_unary_ufuncs.py]: Addressed.

Comment (Skylion007) [reply test/test_unary_ufuncs.py]: Why not have other backend also test fp32 out of curiosity?

Comment (colalb1) [reply test/test_unary_ufuncs.py]: Must have missed that. I will add it.

File: test/test_unary_ufuncs.py [hunk 10]
```diff
@@ -1207,6 +1208,40 @@ def test_sinc(self, device, dtype):
         )
         gradcheck(torch.sinc, a)
 
+    # The order-1 Bessel gradients are indeterminate at x = 0 as written: 0/0 for
+    # j1/i1 (limit 1/2) and (-inf) - (-inf) for y1 (limit +inf). OpInfo sample
+    # generation does not reliably emit exact zeros, and y1's domain floors samples
+    # away from 0, so these special-cased values are only covered here.
+    @dtypes(torch.float32, torch.double)
+    @dtypesIfMPS(torch.float32)
```
Comment (Skylion007) [inline test/test_unary_ufuncs.py]: Actually one other slight nit. MPS support torch half and bfloat16 (but we never added the overload to CUDA). May want to exercise it, doesn't really test much else though.

**Comment by Skylion007** [comment]
Sigh you need to modify opinfo args to specify dtypes for backward only since they don't match exactly looks like?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 11]
```diff
@@ -6162,46 +6162,85 @@ Tensor i1_backward(
...
-               self_is_not_tiny, gradx, at::scalar_tensor(0.5, self.options()));
+        at::where(self_is_tiny, at::scalar_tensor(0.5, self.options()), gradx);
   });
 }
 
+Tensor bessel_j1_backward(
+    const Tensor& grad,
+    const Tensor& self,
+    const Tensor& result) {
+  return AT_DISPATCH_FLOATING_TYPES(
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Yeah, this doesn't dispatch to float16 or bfloat16

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: Ah ok. I'll fix it shortly.

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp]: @Skylion007 Addressed.

**Comment by Skylion007** [comment]
Fix the remaining tests.

**Comment by colalb1** [comment]
> Fix the remaining tests.

@Skylion007 
The three tests that are failing are: `test_output_grad_match_special_i1e_mps_float16`, `test_output_grad_match_special_i1_mps_float16`, `test_comprehensive_special_i1{,e}_cuda_float16` (grad).

The order-1 derivatives have the form `f0(x) - f1(x)/x`, where the terms nearly cancel. With `C = |term| / |result|`, the fp16 backward has two half-ULP error sources, both amplified by `C`: intermediate rounding (every ATen op writes an fp16 tensor), and the saved forward `result`, which arrives already rounded. $C\approx 16$ at the failing i1e sample and is unbounded near the zeros of the derivative, leaving the backends ~1.7% apart against an fp16 rtol of 1e-3.

A few options to correct this:
- Promote to float and recompute `f1(x)` at float, half/bfloat16 only. This removes both sources but costs one extra special-function eval and ignores the saved `result`.
- Promote to float only. Smaller change, no extra eval; leaves the amplified `result` error, so j1/y1 stay wrong and i1 passes.
- Drop half/bfloat16 backward. There is no numerics risk but removes the half support you asked for. 

I'd like to have half-precision support, but I'd prefer not to push something unstable that will raise problems with further development. I recommend the first option; I wanted to run this by you before changing it so we can close this issue with no further changes. Please let me know your thoughts when you get the chance.

Thank you.

**Comment by Skylion007** [comment]
Nah, you can just raise the tolerance of the fp16 using a dtype override probably

**Comment by Skylion007** [comment]
@claude show you can add the dtype precision override to the opinfo and other tests here for the new functionality

**Comment by colalb1** [comment]
@Skylion007 It seems that the tests are passing now. The one failure that Dr CI marked as "probably related" is due to a timeout in the test runner.

**Comment by Skylion007** [comment]
"The PR enables autograd support for 8 Bessel functions that previously had supports_autograd=False. This adds new gradient test cases to test_ops in the inductor shard, pushing the already-borderline MI350 ROCm job past its 30-minute timeout."

**Comment by Skylion007** [comment]
@claude best way to fix this?

**Review by Skylion007 — APPROVED** [review]
